llvm/lib/Analysis/CodeMetrics.cpp

```diff
@@ -163,20 +175,34 @@ void CodeMetrics::analyzeBasicBlock(
     if (isa<ExtractElementInst>(I) || I.getType()->isVectorTy())
       ++NumVectorInsts;
 
-    if (I.getType()->isTokenTy() && I.isUsedOutsideOfBlock(BB))
+    if (L && isa<ConvergenceControlInst>(I)) {
+      if (isUsedOutsideOfLoop(I, *L)) {
+        LLVM_DEBUG(dbgs() << I
+                          << "\n  Cannot duplicate a convergence control token "
+                             "used outside the loop.\n");
+        notDuplicatable = true;
```

Comment:
This seems like it's overloading the meaning of "notDuplicatable"; maybe consider splitting the convergence stuff into a separate bit?

Reply:
The meaning of "notDuplicatable" is not overloaded; it's exactly what the name says, right? It's just that now a block may be "not duplicatable" for two different reasons. But we don't really have any clients that need to distinguish why something is notDuplicatable. We could revisit this variable when such a distinction becomes necessary?

Reply:
FunctionSpecialization is currently using CodeMetrics; does it really not care about the distinction?

Reply:
I still think we should get rid of noduplicate entirely. It was a failed attempt to avoid inventing convergence

Reply:
Fixed. `canUnroll()` now separately considers any discovered convergence instead of mixing it with `notDuplicatable`.